libraries/AP_Compass/AP_Compass_IST8310.cpp

```diff
@@ -0,0 +1,176 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+/*
+ * Copyright (C) 2016  Emlid Ltd. All rights reserved.
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * Driver by Georgii Staroselskii, Sep 2016
+ */
+
+#include "AP_Compass_IST8310.h"
+
+#include <AP_HAL/AP_HAL.h>
+
+#include <utility>
+
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_Math/AP_Math.h>
+
+#define WAI_REG 0x0
+#define DEVICE_ID 0x10
+
+#define CNTL1_REG 0xA
+#define SINGLE_MEASUREMENT_MODE 0x1
+#define ODR_100HZ 0x6
+
+#define STAT1_REG 0x2
+#define DATA_RDY 0x1
+
+extern const AP_HAL::HAL &hal;
+
+AP_Compass_Backend *AP_Compass_IST8310::probe(Compass &compass,
+                                             AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    AP_Compass_IST8310 *sensor = new AP_Compass_IST8310(compass, std::move(dev));
+    if (!sensor || !sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+AP_Compass_IST8310::AP_Compass_IST8310(Compass &compass,
+                                     AP_HAL::OwnPtr<AP_HAL::Device> dev)
+    : AP_Compass_Backend(compass)
+    , _dev(std::move(dev))
+{
+}
+
+bool AP_Compass_IST8310::init()
+{
+    uint8_t val = 0;
+    bool ret;
+
+    hal.scheduler->suspend_timer_procs();
+
+    if (!_dev->get_semaphore()->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        hal.console->printf("IST8310: Unable to get bus semaphore\n");
+        goto fail_sem;
+    }
+
+    ret = _dev->read_registers(WAI_REG, &val, 1);
+    if (!ret) {
+        goto bus_error;
+    }
+    if (val != DEVICE_ID) {
+        hal.console->printf("IST8310: Wrong id\n");
```

Comment:
we are moving away from these "wrong id", particularly for sensors that can be connected on external buses. This is the sensor that comes on the Intel Aero RTF, so I know it can be an external sensor...  Reason is that we want to call probe() possibly on more sensors than we will have on any board. Since we added support for "device id" on the Device API it's safe to do so and know when a sensor appeared or disappeared and know what are the offsets for each sensor.

Summary: you should drop the printf here and check recent changes to current drivers. You can use HMC5843 as an example.